---
title: "GroupWork3"
output: html_document
date: "2022-12-21"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Data

The file `Genus_otu_table.txt` is tab delimited file. Contains raw count data of 16S sequencing of fecal samples from healthy donors and patients affected by clostridium difficile, summarized at genus level

```{r}
filename="Genus_otu_table.txt"
DATA=read.table(filename,header = TRUE, sep = "\t")
#DATA = DATA[1:4, 1:3]
```

#Packages

```{r}
if(!require(compositions)) install.packages(compositions)
library(compositions)
o2=clr(DATA)
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("metagenomeSeq")
library(metagenomeSeq)
```

#PART A

Chose which type(s) of **data transformation** (`clr`, `alr`) **normalization** (`GMPR`, `CSS`, `TMM`) **imputation** (`mbImoute`) you want to perform on data

## Data Transofrmation
$clr(x)=\left(\ln x_i-\frac 1 D \sum_{j=1}^D\ln x_j\right)_i=\ln\left(\frac{x_i}{\Pi_{j=1}^Dx_j}\right)$

```{r}
clr <- function (DATA){
  data= apply(
    DATA
    , MARGIN = 2 #by column 
    , FUN = ( function (x) (x / (prod(x[x != 0]) ^ (1 / length(x[x != 0] ))))))
  notzero = data > 0 & !is.na(data) 
  data = ifelse(notzero,log2(data), data)
  return(data)
}
data_transformed_crl = clr(DATA)
type(data_transformed_crl)
```

```{r}
alr <- function (DATA){
  not_zero = DATA > 0 & !is.na(DATA)
  DATA1 = ifelse(not_zero, 0, 1)
  num_zeros = rowSums(DATA1)
  lower_row = sort(unlist(num_zeros))
  index = which(num_zeros == lower_row[1]) [1]
  DATA[DATA == 0] = 1 #replace 0's with 1's
  row = t(DATA[index,]) #transpose it to do row / colum calc in apply
  
  #problem we haven't a taxa present in all the sample we have to replace 0's with 1's. Also, I am going to     use as reference the one with lower number of 0's 
  for (i in 1:nrow(DATA)){
    DATA[i,] = log2(DATA[i,] / row)
  }
  
  return(DATA)
}

data_transformed_arl = alr(DATA)
data_transformed_arl = as.matrix(data_transformed_arl)
```

##Normalization
```{r}
metaSeqObject = newMRexperiment(data_transformed_crl) #samples on the column and row has to be the feature
metaSeqObject_CSS  = cumNorm( metaSeqObject , p = cumNormStatFast(metaSeqObject) )
OTU_read_count_CSS = data.frame( MRcounts(metaSeqObject_CSS, norm = FALSE, log = TRUE))
```
```{r}
GMPR <- function (DATA){
  
  r_j_k <- vector()
  for (j in 1:ncol(DATA)){
    for (k in 1:ncol(DATA)){
      if(j != k){
        r_j_k <- append(r_j_k , median(DATA[,j] / DATA[,k], na.rm = TRUE))
      }
    }
  }
  
  s_j <- vector()
  value = 0
  
  for(i in length(r_j_k)){
    print(i)
    if (value == 0){
      value =  r_j_k[i]
    }else{
      value = value * r_j_k[i]  
    }
    
    if((i %% (dim(DATA)[2] - 1)) == 0){
      s_j <- append(s_j, value)
      value = 0
    }
  }
  return(s_j)
}

s_j = GMPR(DATA)
```

# Part b

Run **DA analysis** using `Aldex` and `Ancom` and **compare results**

# Part c

Based on the results obtained in [Part b](Part%20b), **select a set of DA taxa** and, based only on those taxa, **run `NMDS`, `t-sne` and `UMAP`** to project healthy donors and the patients on 2 dimension.

> Are These plots equal to the ones obtained using all taxa instead of using only the DA?
