---
title: "GroupWork3"
output: html_document
date: "2022-12-21"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Data

The file `Genus_otu_table.txt` is tab delimited file. Contains raw count data of 16S sequencing of fecal samples from healthy donors and patients affected by clostridium difficile, summarized at genus level

```{r}
filename="Genus_otu_table.txt"
DATA=as.matrix(read.table(filename,header = TRUE, sep = "\t"))
#DATA = DATA[1:4, 1:3]

paste("Sparsity =", round(sum(DATA==0)/length(DATA)*100, digits = 1))
```

# Packages

```{r}
if (!requireNamespace("devtools", quietly = TRUE))  install.packages("devtools")
library(devtools)

if(!require(compositions)) install.packages(compositions)
if (!requireNamespace("BiocManager", quietly = TRUE))  install.packages("BiocManager")
BiocManager::install("metagenomeSeq")
if (!requireNamespace("GUniFrac", quietly = TRUE))  install.packages("GUniFrac")
if (!requireNamespace("glmnet", quietly = TRUE))  install.packages("glmnet")
if (!requireNamespace("Matrix", quietly = TRUE))  install.packages("Matrix")
if (!requireNamespace("mbImpute", quietly = TRUE))  install_github("ruochenj/mbImpute/mbImpute R package")

library(GUniFrac)
library(metagenomeSeq)
library(glmnet)
library(Matrix)
library(compositions)
library(mbImpute)
library(devtools)
```

# PART A

Chose which type(s) of **data transformation** (`clr`, `alr`) **normalization** (`GMPR`, `CSS`, `TMM`) **imputation** (`mbImoute`) you want to perform on data

## Data Transofrmation

$clr(x)=\left(\ln x_i-\frac 1 D \sum_{j=1}^D\ln x_j\right)_i=\ln\left(\frac{x_i}{\Pi_{j=1}^Dx_j}\right)$

```{r}
my.clr <- function (data){
  np<-dim(data)[2] #number of samples
  nt<-dim(data)[1] #number of taxa
  out<-data 
  for (i in (1:np)) {
    den<-(prod(data[,i]))^(1/nt) #geometric mean of column i
    out[,i]<-log2(data[,i]/den) #clr transformation of column i
  } 
  out
}
data.psc_myclr=my.clr(DATA+1)
data_myclr=my.clr(DATA)

data.psc_clr=as.matrix(clr(DATA+1))
data_clr=as.matrix(clr(DATA))

head(DATA[,1:5])
head(data.psc_myclr[,1:5])
head(data.psc_clr[,1:5])
head(data_myclr[,1:5])
head(data_clr[,1:5])
```

```{r}
my.alr <- function (data){
  # select reference | 
  nvm = is.NMV(data)
  data = ifelse(nvm, data, 0)

  num.zero=apply(data==0,MARGIN=1,FUN=sum)
  r= which.min(num.zero)
  cat(sprintf("Reference taxa is the %d (%s) with %d zero \n\n",r, rownames(DATA)[r],  num.zero[r])) 
  
  out=data
  #problem we haven't a taxa present in all the sample we have to replace 0's with 1's. Also, I am going to     use as reference the one with lower number of 0's 
  for (i in 1:ncol(data)){
    if(data[r,i]==0)
      out[,i] = log2(data[,i] / 1)
    else
      out[,i] = log2(data[,i] / DATA[r,i])
  }
  return(out)
}

data_myalr = my.alr(DATA)
data_alr = alr(DATA)

head(DATA[,1:5])
head(data_myalr[,1:5])
head(data_alr[,1:5])
```

## Normalization

```{r}
metaSeqObject = newMRexperiment(data_transformed_crl) #samples on the column and row has to be the feature
metaSeqObject_CSS  = cumNorm( metaSeqObject , p = cumNormStatFast(metaSeqObject) )
OTU_read_count_CSS = data.frame( MRcounts(metaSeqObject_CSS, norm = FALSE, log = TRUE))
```

## GMPR

```{r}
# GMPR expect samples on rows... we need to transpose the count matrix
GMPR_factors<- GMPR(OTUmatrix = t(DATA), min_ct = 2, intersect_no = 4) #see help for parameters meaning
data_gmpr<- t(t(DATA)/GMPR_factors)

GMPR_factors[1:10]
head(DATA[,1:10])
head(data_gmpr[,1:10])
```

# Zero Imputation

## mbImpute

```{r}
label_samples<-read.table("metadata_table.txt",sep="\t",header=T)
# mbImpute expect samples on rows... we need to transpose the count matrix
imp_count_mats <- mbImpute(condition = label_samples$DiseaseState, otu_tab = t(DATA), unnormalized = T)
#mbImpute returns a list three imputed OTU matrices. 
#imp_count_mat_lognorm: imputed normalized and log transformed matrix. 
#imp_count_mat_norm : imputed normalized count matrix with
#library size of each sample equal to 10^6. 
#imp_count_mat_origlibsize: imputed countmatrix at the original library size.

imp_count_mat <-t(imp_count_mats[[3]])
```

## mbImpute with normalized data

```{r}
label_samples<-read.table("metadata_table.txt",sep="\t",header=T)
# mbImpute expect samples on rows... we need to transpose the count matrix
imp_count_mats_gmpr <- mbImpute(condition = label_samples$DiseaseState, otu_tab = t(data_gmpr), unnormalized = F)
#mbImpute returns a list three imputed OTU matrices. 
#imp_count_mat_lognorm: imputed normalized and log transformed matrix. 
#imp_count_mat_norm : imputed normalized count matrix with
#library size of each sample equal to 10^6. 
#imp_count_mat_origlibsize: imputed countmatrix at the original library size.

imp_count_mat_gmpr <-t(imp_count_mats_gmpt[[3]])
```

# Part b

Run **DA analysis** using `Aldex` and `Ancom` and **compare results**

# Part c

Based on the results obtained in [Part b](Part%20b), **select a set of DA taxa** and, based only on those taxa, **run `NMDS`, `t-sne` and `UMAP`** to project healthy donors and the patients on 2 dimension.

> Are These plots equal to the ones obtained using all taxa instead of using only the DA?
