---
title: "2-project"
date: "2022-11-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# MvAplot
Takes in input 
* exprData: rows <-> genes; columns <-> samples; cells contains the number of transcripts of each gene found
* pdffilename: filename to be used in the cration of pdf file containing all MvA plots (1 vs i for i in (2:N))

```{r}
MvAplot <- function(exprData, 
                    pdffilename){
  ngenes=dim(exprData)[1]
  nsamples=dim(exprData)[2]
  pdf(pdffilename)
  par(mfrow=c(2,nsamples/2),
      xlab=)
  
  for(i in 2:nsamples){
    M=log2(exprData[,1]/exprData[,i])
    A=0.5*log2(exprData[,1]*exprData[,i])
    plot(A, M)
  }
  
  dev.off()
}
```

##TMMnorm
1. Takes as input 
  • exprData: a numeric data matrix that is supposed to have     
    the same format of the raw transcript 
    data provided in stem (i.e. genes’IDs on rows and            
    subjects’IDs on columns)
  • annot: a data frame that is supposed to have the same        
    format of the annotation data provided in 
    stem
  • Atrim: a vector of 2 elements indicating the lower and       
    upper thresholds to trim the most extreme 
    values of A (where A is the average in log2 scale) 
  • Mtrim: a number between 0 and 1 indicating the fraction    
    of observations to be deleted from each 
    end (positive and negative values) of the sorted vector M    
    before calculating the mean (where Ms
    are the log ratios defined as in the MvA plot
  
2. The function TMMnorm
  • Scale the data by their sequencing depth and multiply by 10^6
  • Calculates the scaling factors SF (with respect to sample 1) 
    by trimming the most extreme values of 
    A and taking the trimmed means of M values (suggestion:
    use the R function mean)
  • Normalizes the data by their scaling factors SF with         
    respect to sample 1
  • Scale the genes (in the original scale, not in log scale)     
    by their length and multiply by 10^3
  • Returns a list of 2 elements: i) the vector of the           
    scaling factors; ii) the normalized matrix        
    (in the original scale, not in log scale)
####Suggestion: Generates a .pdf file with MvA plots before      and after the normalization to check your work (use the function MvAplot)
```{r}

TMMnorm <- function(exprData = RawTranscript,
                    annot = RawTranscript_annot,
                    Atrim = c(0,8),
                    Mtrim = 0.02){
}


```

# Tests
```{r}
exprData=read.table("raw_trascr_count.txt")

MvAplot(data, "MvAPlots.pdf")

```